File: Online_Judge/LeetCode/C++/98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
//9ms
class Solution {
public:
    vector<int>itv;
    bool isValidBST(TreeNode* root) {
        inorder(root);
        unsigned int size = itv.size();
        for (unsigned int i = 1;i < size;i++){
            if (itv[i] <= itv[i - 1]){
                return false;
            }
        }
        return true;
    }

    void inorder(TreeNode* root){
        if (root != NULL){
            inorder(root->left);
            itv.push_back(root->val);
            inorder(root->right);
        }
    }
};

//9ms
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        TreeNode *prev = nullptr,*curr = root;
        stack<TreeNode*> store;

        while(!store.empty()||curr){
            while(curr){
                store.push(curr);
                curr=curr->left;
            }
            TreeNode* top = store.top();
            store.pop();
            if(!prev){
                prev = top;
            }else{
                if(prev->val>=top->val)
                    return false;
                prev = top;
            }
            curr=top->right;
        }
        return true;

    }
};

File: LeetCode/Java/98. Validate Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    //4ms
    private List<Integer> list= new ArrayList<Integer>();
    public boolean isValidBST(TreeNode root) {
        inorder(root);
        int size = list.size();
        for (int i = 1;i < size;i++){
            if (list.get(i) <= list.get(i - 1))
                return  false;
        }
        return true;
    }
    void inorder(TreeNode root){
        if (root != null){
            inorder(root.left);
            list.add(root.val);
            inorder(root.right);
        }
    }
}